src/mlpack/methods/ann/layer/threshold_fn.hpp

```diff
@@ -0,0 +1,131 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_THRESHOLD_HPP
+#define MLPACK_METHODS_ANN_LAYER_THRESHOLD_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Threshold activation function, defined by
+ *
+ * @f{eqnarray*}{
+ * f(x) &=& \left\{
+ *   \begin{array}{lr}
+ *     x & : x > threshold \\
+ *     value & : x \le threshold
+ *   \end{array}
+ * \right.
+ * @f}
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class Threshold
+{
+ public:
+  /**
+   * Create the Threshold object using the specified parameters.
+   * The non zero gradient region can be adjusted by specifying the parameter
+   * threshold. Default (threshold = 0.0)
+   * The value when x < threshold can be adjusted by specifying the parameter
+   * value. Default (value = 0.0)
+   *
+   * @param threshold Non zero gradient region threshold
+   * @param value Value when x < threshold
+   */
+  Threshold(const double threshold = 0.0,
+            const double value = 0.0);
+
+  /**
+   * Ordinary feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward through f.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename InputType, typename OutputType>
+  void Forward(const InputType& input, OutputType& output);
+
+  /**
+   * Ordinary feed backward pass of a neural network, calculating the function
+   * f(x) by propagating x backwards through f. Using the results from the feed
+   * forward pass.
+   *
+   * @param input The propagated input activation.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename DataType>
+  void Backward(const DataType& input, const DataType& gy, DataType& g);
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the non zero gradient region threshold.
+  double const& ThresholdVar() const { return threshold; }
+  //! Modify the non zero gradient region threshold.
+  double& ThresholdVar() { return threshold; }
```

Comment:
```suggestion
  double const& Threshold() const { return threshold; }
  //! Modify the non zero gradient region threshold.
  double& Threshold() { return threshold; }
```

Reply:
Will it not conflict with class name

Reply:
@zoq

Reply:
Ohh, you are right, in this case let's keep the method name.